nni/algorithms/compression/v2/pytorch/pruning/tools/rl_env/channel_pruning_env.py

```diff
@@ -0,0 +1,585 @@
+
+    def step_after(self, action, preserve_idx):
+        _, action, d_prime, preserve_idx = self.prune_kernel(self.prunable_idx[self.cur_ind], action, preserve_idx)
+        if not self.visited[self.cur_ind]:
+            for group in self.shared_idx:
+                if self.cur_ind in group:  # set the shared ones
+                    for g_idx in group:
+                        self.strategy_dict[self.prunable_idx[g_idx]][0] = action
+                        self.strategy_dict[self.prunable_idx[g_idx - 1]][1] = action
+                        self.visited[g_idx] = True
+                        self.index_buffer[g_idx] = preserve_idx.copy()
+
+        self.strategy.append(action)  # save action to strategy
+        self.d_prime_list.append(d_prime)
+
+        self.strategy_dict[self.prunable_idx[self.cur_ind]][0] = action
+        if self.cur_ind > 0:
+            self.strategy_dict[self.prunable_idx[self.cur_ind - 1]][1] = action
+
+        # all the actions are made
+        if self._is_final_layer():
+            assert len(self.strategy) == len(self.prunable_idx)
+            current_flops = self._cur_flops()
+            compress_ratio = current_flops * 1. / self.org_flops
+            reward = self.cur_reward
+            info_set = {'compress_ratio': compress_ratio, 'reward': reward, 'strategy': self.strategy.copy(), 'config_list': copy.deepcopy(self.temp_config_list)}
+
+            if reward > self.best_reward and self.ini:
+                self.best_reward = reward
+                self.best_strategy = self.strategy.copy()
+                self.best_d_prime_list = self.d_prime_list.copy()
+                _logger.info('New best reward: %.4f, reward: %.4f, compress: %.4f', self.best_reward, reward, compress_ratio)
+                _logger.info('New best policy: %s', self.best_strategy)
+                _logger.info('New best d primes: %s', self.best_d_prime_list)
+            obs = self.layer_embedding[self.cur_ind, :].copy()  # actually the same as the last state
+            done = True
+            self.temp_config_list = []
+            self.ini = True
+            return obs, reward, done, info_set
+        
+        reward = 0
+        done = False
+        self.visited[self.cur_ind] = True  # set to visited
+        self.cur_ind += 1  # the index of next layer
+        info_set = {'compress_ratio': self._cur_flops() * 1. / self.org_flops, 'reward': reward, 'strategy': self.strategy.copy(), 'config_list': copy.deepcopy(self.temp_config_list)}
+        # build next state (in-place modify)
+        self.layer_embedding[self.cur_ind][-3] = self._cur_reduced() * 1. / self.org_flops  # reduced
+        self.layer_embedding[self.cur_ind][-2] = sum(self.flops_list[self.cur_ind + 1:]) * 1. / self.org_flops  # rest
+        self.layer_embedding[self.cur_ind][-1] = self.strategy[-1]  # last action
+        obs = self.layer_embedding[self.cur_ind, :].copy()
+
+        return obs, reward, done, info_set
+
+    def step(self, act):
```

Comment:
this function can be removed

Reply:
It is remained for backup, now it could be removed.